# MIT License
#
# Copyright (C) 2025 Advanced Micro Devices, Inc. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9a988c5d2..f166fb129 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,10 +40,12 @@ if(FAISS_ENABLE_CUVS)
   include(rapids-cuda)
   include(rapids-export)
   include(rapids-find)
-
-  rapids_cuda_init_architectures(faiss)
-  rapids_cuda_init_architectures(pyfaiss)
-  rapids_cuda_init_architectures(faiss_c_library)
+
+  if(NOT FAISS_ENABLE_ROCM)
+    rapids_cuda_init_architectures(faiss)
+    rapids_cuda_init_architectures(pyfaiss)
+    rapids_cuda_init_architectures(faiss_c_library)
+  endif()
 endif()

 project(faiss
diff --git a/faiss/gpu/CMakeLists.txt b/faiss/gpu/CMakeLists.txt
index 0051f047f..e4470e447 100644
--- a/faiss/gpu/CMakeLists.txt
+++ b/faiss/gpu/CMakeLists.txt
@@ -256,6 +256,11 @@ if(FAISS_ENABLE_CUVS)
           impl/CuvsIVFFlat.cu
           impl/CuvsIVFPQ.cu
           utils/CuvsUtils.cu)
+
+  if (FAISS_ENABLE_ROCM)
+    list(TRANSFORM FAISS_GPU_SRC REPLACE cu$ hip)
+  endif()
+
 endif()

 add_library(faiss_gpu_objs OBJECT ${FAISS_GPU_SRC})
@@ -324,7 +329,7 @@ foreach(header ${FAISS_GPU_HEADERS})
 endforeach()

 if (FAISS_ENABLE_ROCM)
-  target_link_libraries(faiss_gpu_objs PRIVATE hip::host roc::hipblas)
+  target_link_libraries(faiss_gpu_objs PRIVATE hip::host roc::hipblas $<$<BOOL:${FAISS_ENABLE_CUVS}>:cuvs::cuvs>)
   target_compile_options(faiss_gpu_objs PRIVATE)
 else()
   # Prepares a host linker script and enables host linker to support
diff --git a/faiss/gpu/utils/DeviceDefs.cuh b/faiss/gpu/utils/DeviceDefs.cuh
index dc4c469a1..944211da3 100644
--- a/faiss/gpu/utils/DeviceDefs.cuh
+++ b/faiss/gpu/utils/DeviceDefs.cuh
@@ -9,16 +9,17 @@

 #include <cuda.h>

+#ifdef USE_AMD_ROCM
+#include <rocprim/rocprim.hpp>
+#endif
+
 namespace faiss {
 namespace gpu {

 #ifdef USE_AMD_ROCM

-#if __AMDGCN_WAVEFRONT_SIZE == 32u
-constexpr int kWarpSize = 32;
-#else
-constexpr int kWarpSize = 64;
-#endif
+__device__ constexpr int kWarpSize = rocprim::arch::wavefront::max_size();
+

 // This is a memory barrier for intra-warp writes to shared memory.
 __forceinline__ __device__ void warpFence() {
diff --git a/faiss/gpu/utils/MatrixMult-inl.cuh b/faiss/gpu/utils/MatrixMult-inl.cuh
index b28ea56da..ae035715f 100644
--- a/faiss/gpu/utils/MatrixMult-inl.cuh
+++ b/faiss/gpu/utils/MatrixMult-inl.cuh
@@ -24,17 +24,17 @@ struct GetCudaType;

 template <>
 struct GetCudaType<float> {
-    static constexpr hipblasDatatype_t Type = HIPBLAS_R_32F;
+    static constexpr hipDataType Type = HIPBLAS_R_32F;
 };

 template <>
 struct GetCudaType<half> {
-    static constexpr hipblasDatatype_t Type = HIPBLAS_R_16F;
+    static constexpr hipDataType Type = HIPBLAS_R_16F;
 };

 template <>
 struct GetCudaType<__hip_bfloat16> {
-    static constexpr hipblasDatatype_t Type = HIPBLAS_R_16B;
+    static constexpr hipDataType Type = HIPBLAS_R_16B;
 };

 #else
@@ -94,7 +94,7 @@ cublasStatus_t rawGemm(
             C,
             HIPBLAS_R_32F,
             ldc,
-            HIPBLAS_R_32F,
+            HIPBLAS_COMPUTE_32F,
             HIPBLAS_GEMM_DEFAULT);
 #else

@@ -199,7 +199,7 @@ cublasStatus_t rawBatchGemm(
             ldc,
             strideC,
             batchCount,
-            HIPBLAS_R_32F,
+            HIPBLAS_COMPUTE_32F,
             HIPBLAS_GEMM_DEFAULT);
 #else
     return cublasGemmStridedBatchedEx(
diff --git a/faiss/gpu/impl/CuvsIVFFlat.cu b/faiss/gpu/impl/CuvsIVFFlat.cu
index 2cccee860..c4cb13f48 100644
--- a/faiss/gpu/impl/CuvsIVFFlat.cu
+++ b/faiss/gpu/impl/CuvsIVFFlat.cu
@@ -427,13 +427,11 @@ void CuvsIVFFlat::copyInvertedListsFrom(const InvertedLists* ivf) {
     // Precompute the centers vector norms for L2Expanded distance
     if (this->metric_ == faiss::METRIC_L2) {
         cuvs_index->allocate_center_norms(raft_handle);
-        raft::linalg::rowNorm(
+        raft::linalg::rowNorm<raft::linalg::L2Norm, true>(
                 cuvs_index->center_norms().value().data_handle(),
                 cuvs_index->centers().data_handle(),
                 cuvs_index->dim(),
                 (uint32_t)nlist,
-                raft::linalg::L2Norm,
-                true,
                 raft_handle.get_stream());
     }
 }
